core/domain/stats_jobs_one_off.py

```diff
@@ -28,14 +34,248 @@
+            yield (item.exp_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION,
+                       'version': item.exp_version,
+                       'state_name': item.state_name
+                   })
+        elif isinstance(
+                item, stats_models.CompleteExplorationEventLogEntryModel):
+            yield (item.exploration_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_COMPLETE_EXPLORATION,
+                       'version': item.exploration_version,
+                       'state_name': item.state_name
+                   })
+
+
+    @staticmethod
+    def reduce(exp_id, values):
+        values = map(ast.literal_eval, values)
+        values = sorted(values, key=lambda x: x['version'])
+        versions = [x['version'] for x in values]
+        # Get a copy of the uncorrupted *_v1 statistics
+        old_stats = stats_services.get_multiple_exploration_stats_by_version(
+            exp_id, versions)
+        # Get list of snapshot models for each version of the exploration
+        snapshots_by_version = (
+            exp_models.ExplorationModel.get_snapshots_metadata(
+                exp_id, versions))
+
+        exp_stats_dicts = []
+        for version, events in itertools.groupby(values,
+                                                 key=lambda x: x['version']):
+            v1_stats = old_stats[version-1]
+            if version == 1:
+                # Reset the possibly corrupted stats
+                v1_stats.num_starts_v2 = 0
+                v1_stats.num_completions_v2 = 0
+                v1_stats.num_starts_v2 = 0
+                for state_stats in v1_stats.state_stats_mapping.values():
+                    state_stats.total_answers_count_v2 = 0
+                    state_stats.useful_feedback_count_v2 = 0
+                    state_stats.total_hit_count_v2 = 0
+                    state_stats.first_hit_count_v2 = 0
+                    state_stats.num_times_solution_viewed_v2 = 0
+                    state_stats.num_completions_v2 = 0
+                exp_stats_dict = RecomputeStatistics._recompute_statistics(
+                    exp_id, version, events, v1_stats.to_dict())
+                exp_stats_dicts.append(exp_stats_dict)
+            else:
+                change_list = snapshots_by_version[version - 1]['commit_cmds']
+                # Copy recomputed v2 events from previous version
+                prev_version = copy.deepcopy(exp_stats_dicts[-1])
+
+                # Copy uncorrupt v1 stats
+                prev_version['num_starts_v1'] = v1_stats.num_starts_v1
+                prev_version['num_completions_v1'] = v1_stats.num_completions_v1
+                prev_version['num_actual_starts_v1'] = (
+                    v1_stats.num_actual_starts_v1)
+                state_stats_mapping = prev_version['state_stats_mapping']
+                for state in state_stats_mapping.keys():
```

Comment:
You can drop ".keys()" -- more idiomatic (and possibly faster since you don't need to construct the extra list).

Reply:
Done.